---
description: –ß–∏—Å—Ç—ã–µ —Ñ—É–Ω–∫—Ü–∏–∏ –≤—ã–ø–æ–ª–Ω—è—é—Ç —Ç–æ–ª—å–∫–æ –≤—ã—á–∏—Å–ª–µ–Ω–∏—è –∏ –Ω–∏—á–µ–≥–æ –±–æ–ª—å—à–µ. –≠—Ç–æ –æ–±–ª–µ–≥—á–∞–µ—Ç –ø–æ–Ω–∏–º–∞–Ω–∏–µ –∏ –æ—Ç–ª–∞–¥–∫—É –∫–æ–¥–∞, –∞ —Ç–∞–∫–∂–µ –ø–æ–∑–≤–æ–ª—è–µ—Ç React –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏ –æ–ø—Ç–∏–º–∏–∑–∏—Ä–æ–≤–∞—Ç—å –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –∏ —Ö—É–∫–∏
---

# –ö–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –∏ —Ö—É–∫–∏ –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å —á–∏—Å—Ç—ã–º–∏

<big>
–ß–∏—Å—Ç—ã–µ —Ñ—É–Ω–∫—Ü–∏–∏ –≤—ã–ø–æ–ª–Ω—è—é—Ç —Ç–æ–ª—å–∫–æ –≤—ã—á–∏—Å–ª–µ–Ω–∏—è –∏ –Ω–∏—á–µ–≥–æ –±–æ–ª—å—à–µ. –≠—Ç–æ –æ–±–ª–µ–≥—á–∞–µ—Ç –ø–æ–Ω–∏–º–∞–Ω–∏–µ –∏ –æ—Ç–ª–∞–¥–∫—É –∫–æ–¥–∞, –∞ —Ç–∞–∫–∂–µ –ø–æ–∑–≤–æ–ª—è–µ—Ç React –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏ –æ–ø—Ç–∏–º–∏–∑–∏—Ä–æ–≤–∞—Ç—å –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –∏ —Ö—É–∫–∏.
</big>

!!!note "–ß–∏—Å—Ç–æ—Ç–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤"

    –≠—Ç–∞ —Å–ø—Ä–∞–≤–æ—á–Ω–∞—è —Å—Ç—Ä–∞–Ω–∏—Ü–∞ –æ—Ö–≤–∞—Ç—ã–≤–∞–µ—Ç —Ä–∞—Å—à–∏—Ä–µ–Ω–Ω—ã–µ —Ç–µ–º—ã –∏ —Ç—Ä–µ–±—É–µ—Ç –∑–Ω–∞–∫–æ–º—Å—Ç–≤–∞ —Å –∫–æ–Ω—Ü–µ–ø—Ü–∏—è–º–∏, —Ä–∞—Å—Å–º–æ—Ç—Ä–µ–Ω–Ω—ã–º–∏ –Ω–∞ —Å—Ç—Ä–∞–Ω–∏—Ü–µ [¬´–ß–∏—Å—Ç–æ—Ç–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤¬ª](../../learn/keeping-components-pure.md).

## –ü–æ—á–µ–º—É —á–∏—Å—Ç–æ—Ç–∞ –∏–º–µ–µ—Ç –∑–Ω–∞—á–µ–Ω–∏–µ? {#why-does-purity-matter}

–û–¥–Ω–∞ –∏–∑ –∫–ª—é—á–µ–≤—ã—Ö –∫–æ–Ω—Ü–µ–ø—Ü–∏–π, –∫–æ—Ç–æ—Ä–∞—è –¥–µ–ª–∞–µ—Ç React _React_, - —ç—Ç–æ _—á–∏—Å—Ç–æ—Ç–∞_. –ß–∏—Å—Ç—ã–π –∫–æ–º–ø–æ–Ω–µ–Ω—Ç –∏–ª–∏ —Ö—É–∫ - —ç—Ç–æ —Ç–æ—Ç, –∫–æ—Ç–æ—Ä—ã–π:

-   **–ò–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–µ–Ω** - –í—ã [–≤—Å–µ–≥–¥–∞ –ø–æ–ª—É—á–∞–µ—Ç–µ –æ–¥–∏–Ω –∏ —Ç–æ—Ç –∂–µ —Ä–µ–∑—É–ª—å—Ç–∞—Ç –∫–∞–∂–¥—ã–π —Ä–∞–∑](../../learn/keeping-components-pure.md#purity-components-as-formulas), –∫–æ–≥–¥–∞ –∑–∞–ø—É—Å–∫–∞–µ—Ç–µ –µ–≥–æ —Å —Ç–µ–º–∏ –∂–µ –≤—Ö–æ–¥–Ω—ã–º–∏ –¥–∞–Ω–Ω—ã–º–∏ - props, state, context –¥–ª—è –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–Ω—ã—Ö –≤—Ö–æ–¥–æ–≤; –∏ –∞—Ä–≥—É–º–µ–Ω—Ç—ã –¥–ª—è –≤—Ö–æ–¥–æ–≤ —Ö—É–∫–∞.
-   **–ù–µ –∏–º–µ–µ—Ç –ø–æ–±–æ—á–Ω—ã—Ö —ç—Ñ—Ñ–µ–∫—Ç–æ–≤ –≤ —Ä–µ–Ω–¥–µ—Ä–µ** - –ö–æ–¥ —Å –ø–æ–±–æ—á–Ω—ã–º–∏ —ç—Ñ—Ñ–µ–∫—Ç–∞–º–∏ –¥–æ–ª–∂–µ–Ω –≤—ã–ø–æ–ª–Ω—è—Ç—å—Å—è [**–æ—Ç–¥–µ–ª—å–Ω–æ –æ—Ç —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞**](#how-does-react-run-your-code). –ù–∞–ø—Ä–∏–º–µ—Ä, –∫–∞–∫ [–æ–±—Ä–∞–±–æ—Ç—á–∏–∫ —Å–æ–±—ã—Ç–∏–π](../../learn/responding-to-events.md) - –∫–æ–≥–¥–∞ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å –≤–∑–∞–∏–º–æ–¥–µ–π—Å—Ç–≤—É–µ—Ç —Å –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–º –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–æ–º –∏ –∑–∞—Å—Ç–∞–≤–ª—è–µ—Ç –µ–≥–æ –æ–±–Ω–æ–≤–ª—è—Ç—å—Å—è; –∏–ª–∏ –∫–∞–∫ [—ç—Ñ—Ñ–µ–∫—Ç](../react/useEffect.md) - –∫–æ—Ç–æ—Ä—ã–π –∑–∞–ø—É—Å–∫–∞–µ—Ç—Å—è –ø–æ—Å–ª–µ —Ä–µ–Ω–¥–µ—Ä–∞.
-   **–ù–µ –º—É—Ç–∏—Ä—É–µ—Ç –Ω–µ–ª–æ–∫–∞–ª—å–Ω—ã–µ –∑–Ω–∞—á–µ–Ω–∏—è**: –ö–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –∏ —Ö—É–∫–∏ –¥–æ–ª–∂–Ω—ã [–Ω–∏–∫–æ–≥–¥–∞ –Ω–µ –∏–∑–º–µ–Ω—è—Ç—å –∑–Ω–∞—á–µ–Ω–∏—è, –∫–æ—Ç–æ—Ä—ã–µ –Ω–µ —Å–æ–∑–¥–∞—é—Ç—Å—è –ª–æ–∫–∞–ª—å–Ω–æ](#mutation) –ø—Ä–∏ —Ä–µ–Ω–¥–µ—Ä–µ.

–ö–æ–≥–¥–∞ —Ä–µ–Ω–¥–µ—Ä –æ—Å—Ç–∞–µ—Ç—Å—è —á–∏—Å—Ç—ã–º, React –º–æ–∂–µ—Ç –ø–æ–Ω—è—Ç—å, –∫–∞–∫ —Ä–∞—Å—Å—Ç–∞–≤–∏—Ç—å –ø—Ä–∏–æ—Ä–∏—Ç–µ—Ç—ã, –∫–∞–∫–∏–µ –æ–±–Ω–æ–≤–ª–µ–Ω–∏—è –Ω–∞–∏–±–æ–ª–µ–µ –≤–∞–∂–Ω—ã –¥–ª—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è, —á—Ç–æ–±—ã —É–≤–∏–¥–µ—Ç—å –∏—Ö –ø–µ—Ä–≤—ã–º–∏. –≠—Ç–æ —Å—Ç–∞–Ω–æ–≤–∏—Ç—Å—è –≤–æ–∑–º–æ–∂–Ω—ã–º –±–ª–∞–≥–æ–¥–∞—Ä—è —á–∏—Å—Ç–æ—Ç–µ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞: –ø–æ—Å–∫–æ–ª—å–∫—É –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –Ω–µ –∏–º–µ—é—Ç –ø–æ–±–æ—á–Ω—ã—Ö —ç—Ñ—Ñ–µ–∫—Ç–æ–≤ [–≤ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–µ](#how-does-react-run-your-code), React –º–æ–∂–µ—Ç –ø—Ä–∏–æ—Å—Ç–∞–Ω–æ–≤–∏—Ç—å —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤, –∫–æ—Ç–æ—Ä—ã–µ –Ω–µ —Ç–∞–∫ –≤–∞–∂–Ω–æ –æ–±–Ω–æ–≤–ª—è—Ç—å, –∏ –≤–µ—Ä–Ω—É—Ç—å—Å—è –∫ –Ω–∏–º –ø–æ–∑–∂–µ, –∫–æ–≥–¥–∞ —ç—Ç–æ –±—É–¥–µ—Ç –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ.

–ö–æ–Ω–∫—Ä–µ—Ç–Ω–æ —ç—Ç–æ –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ –ª–æ–≥–∏–∫–∞ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ –º–æ–∂–µ—Ç –±—ã—Ç—å –∑–∞–ø—É—â–µ–Ω–∞ –Ω–µ—Å–∫–æ–ª—å–∫–æ —Ä–∞–∑, —á—Ç–æ –ø–æ–∑–≤–æ–ª—è–µ—Ç React –æ–±–µ—Å–ø–µ—á–∏—Ç—å –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—é –ø—Ä–∏—è—Ç–Ω—ã–π –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–π –æ–ø—ã—Ç. –û–¥–Ω–∞–∫–æ –µ—Å–ª–∏ —É –≤–∞—à–µ–≥–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ –µ—Å—Ç—å –Ω–µ–æ—Ç—Å–ª–µ–∂–∏–≤–∞–µ–º—ã–π –ø–æ–±–æ—á–Ω—ã–π —ç—Ñ—Ñ–µ–∫—Ç - –Ω–∞–ø—Ä–∏–º–µ—Ä, –∏–∑–º–µ–Ω–µ–Ω–∏–µ –∑–Ω–∞—á–µ–Ω–∏—è –≥–ª–æ–±–∞–ª—å–Ω–æ–π –ø–µ—Ä–µ–º–µ–Ω–Ω–æ–π [–≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞](#how-does-react-run-your-code) - –∫–æ–≥–¥–∞ React —Å–Ω–æ–≤–∞ –∑–∞–ø—É—Å—Ç–∏—Ç –≤–∞—à –∫–æ–¥ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞, –≤–∞—à–∏ –ø–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã –±—É–¥—É—Ç –≤—ã–∑–≤–∞–Ω—ã –Ω–µ —Ç–∞–∫, –∫–∞–∫ –≤—ã —Ö–æ—Ç–µ–ª–∏. –≠—Ç–æ —á–∞—Å—Ç–æ –ø—Ä–∏–≤–æ–¥–∏—Ç –∫ –Ω–µ–æ–∂–∏–¥–∞–Ω–Ω—ã–º –æ—à–∏–±–∫–∞–º, –∫–æ—Ç–æ—Ä—ã–µ –º–æ–≥—É—Ç —É—Ö—É–¥—à–∏—Ç—å –≤–æ—Å–ø—Ä–∏—è—Ç–∏–µ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è–º–∏. –í—ã –º–æ–∂–µ—Ç–µ —É–≤–∏–¥–µ—Ç—å [–ø—Ä–∏–º–µ—Ä —ç—Ç–æ–≥–æ –Ω–∞ —Å—Ç—Ä–∞–Ω–∏—Ü–µ Keeping Components Pure](../../learn/keeping-components-pure.md#side-effects-unintended-consequences).

### –ö–∞–∫ React –≤—ã–ø–æ–ª–Ω—è–µ—Ç –≤–∞—à –∫–æ–¥? {#how-does-react-run-your-code}

React —è–≤–ª—è–µ—Ç—Å—è –¥–µ–∫–ª–∞—Ä–∞—Ç–∏–≤–Ω—ã–º: –≤—ã —É–∫–∞–∑—ã–≤–∞–µ—Ç–µ React, —á—Ç–æ –∏–º–µ–Ω–Ω–æ –Ω—É–∂–Ω–æ –æ—Ç–æ–±—Ä–∞–∑–∏—Ç—å, –∞ React —Å–∞–º —Ä–µ—à–∞–µ—Ç, –∫–∞–∫ –ª—É—á—à–µ –ø–æ–∫–∞–∑–∞—Ç—å —ç—Ç–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—é. –î–ª—è —ç—Ç–æ–≥–æ —É React –µ—Å—Ç—å –Ω–µ—Å–∫–æ–ª—å–∫–æ —ç—Ç–∞–ø–æ–≤, –Ω–∞ –∫–æ—Ç–æ—Ä—ã—Ö –æ–Ω –≤—ã–ø–æ–ª–Ω—è–µ—Ç –≤–∞—à –∫–æ–¥. –í–∞–º –Ω–µ –Ω—É–∂–Ω–æ –∑–Ω–∞—Ç—å –æ–±–æ –≤—Å–µ—Ö —ç—Ç–∏—Ö —Ñ–∞–∑–∞—Ö, —á—Ç–æ–±—ã —Ö–æ—Ä–æ—à–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å React. –ù–æ –Ω–∞ –≤—ã—Å–æ–∫–æ–º —É—Ä–æ–≤–Ω–µ –≤—ã –¥–æ–ª–∂–Ω—ã –∑–Ω–∞—Ç—å, –∫–∞–∫–æ–π –∫–æ–¥ –≤—ã–ø–æ–ª–Ω—è–µ—Ç—Å—è –≤ _render_, –∞ –∫–∞–∫–æ–π - –∑–∞ –µ–≥–æ –ø—Ä–µ–¥–µ–ª–∞–º–∏.

–ü–æ–¥ _—Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–æ–º_ –ø–æ–Ω–∏–º–∞–µ—Ç—Å—è –≤—ã—á–∏—Å–ª–µ–Ω–∏–µ —Ç–æ–≥–æ, –∫–∞–∫ –¥–æ–ª–∂–Ω–∞ –≤—ã–≥–ª—è–¥–µ—Ç—å —Å–ª–µ–¥—É—é—â–∞—è –≤–µ—Ä—Å–∏—è –≤–∞—à–µ–≥–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞. –ü–æ—Å–ª–µ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ [—ç—Ñ—Ñ–µ–∫—Ç—ã](../react/useEffect.md) _–ø—Ä–æ–º—ã–≤–∞—é—Ç—Å—è_ (—Ç–æ –µ—Å—Ç—å –≤—ã–ø–æ–ª–Ω—è—é—Ç—Å—è –¥–æ —Ç–µ—Ö –ø–æ—Ä, –ø–æ–∫–∞ –∏—Ö –Ω–µ –æ—Å—Ç–∞–Ω–µ—Ç—Å—è) –∏ –º–æ–≥—É—Ç –æ–±–Ω–æ–≤–∏—Ç—å —Ä–∞—Å—á–µ—Ç—ã, –µ—Å–ª–∏ —ç—Ñ—Ñ–µ–∫—Ç—ã –≤–ª–∏—è—é—Ç –Ω–∞ –º–∞–∫–µ—Ç. React –±–µ—Ä–µ—Ç —ç—Ç–æ—Ç –Ω–æ–≤—ã–π —Ä–∞—Å—á–µ—Ç –∏ —Å—Ä–∞–≤–Ω–∏–≤–∞–µ—Ç –µ–≥–æ —Å —Ä–∞—Å—á–µ—Ç–æ–º, –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–Ω—ã–º –¥–ª—è —Å–æ–∑–¥–∞–Ω–∏—è –ø—Ä–µ–¥—ã–¥—É—â–µ–π –≤–µ—Ä—Å–∏–∏ –≤–∞—à–µ–≥–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞, –∞ –∑–∞—Ç–µ–º _—Ñ–∏–∫—Å–∏—Ä—É–µ—Ç_ —Ç–æ–ª—å–∫–æ –º–∏–Ω–∏–º–∞–ª—å–Ω—ã–µ –∏–∑–º–µ–Ω–µ–Ω–∏—è, –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã–µ –¥–ª—è [DOM](https://developer.mozilla.org/docs/Web/API/Document_Object_Model) (—Ç–æ, —á—Ç–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å –≤–∏–¥–∏—Ç –Ω–∞ —Å–∞–º–æ–º –¥–µ–ª–µ), —á—Ç–æ–±—ã –ø—Ä–∏–≤–µ—Å—Ç–∏ –µ–≥–æ –≤ —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤–∏–µ —Å –ø–æ—Å–ª–µ–¥–Ω–µ–π –≤–µ—Ä—Å–∏–µ–π.

!!!note "–ö–∞–∫ –æ–ø—Ä–µ–¥–µ–ª–∏—Ç—å, –≤—ã–ø–æ–ª–Ω—è–µ—Ç—Å—è –ª–∏ –∫–æ–¥ –ø—Ä–∏ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–µ"

    –û–¥–Ω–∞ –∏–∑ –±—ã—Å—Ç—Ä—ã—Ö —ç–≤—Ä–∏—Å—Ç–∏–∫, –ø–æ–∑–≤–æ–ª—è—é—â–∏—Ö –æ–ø—Ä–µ–¥–µ–ª–∏—Ç—å, –≤—ã–ø–æ–ª–Ω—è–µ—Ç—Å—è –ª–∏ –∫–æ–¥ –≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞, - —ç—Ç–æ –ø–æ—Å–º–æ—Ç—Ä–µ—Ç—å, –≥–¥–µ –æ–Ω –Ω–∞—Ö–æ–¥–∏—Ç—Å—è: –µ—Å–ª–∏ –æ–Ω –Ω–∞–ø–∏—Å–∞–Ω –Ω–∞ –≤–µ—Ä—Ö–Ω–µ–º —É—Ä–æ–≤–Ω–µ, –∫–∞–∫ –≤ –ø—Ä–∏–º–µ—Ä–µ –Ω–∏–∂–µ, —Ç–æ –≤–µ–ª–∏–∫–∞ –≤–µ—Ä–æ—è—Ç–Ω–æ—Å—Ç—å, —á—Ç–æ –æ–Ω –≤—ã–ø–æ–ª–Ω—è–µ—Ç—Å—è –≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞.

    ```js hl_lines="2"
    function Dropdown() {
    	const selectedItems = new Set(); // created during render
    	// ...
    }
    ```

    –û–±—Ä–∞–±–æ—Ç—á–∏–∫–∏ —Å–æ–±—ã—Ç–∏–π –∏ —ç—Ñ—Ñ–µ–∫—Ç—ã –Ω–µ –∑–∞–ø—É—Å–∫–∞—é—Ç—Å—è –ø—Ä–∏ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–µ:

    ```js hl_lines="4-5"
    function Dropdown() {
    	const selectedItems = new Set();
    	const onSelect = (item) => {
    		// this code is in an event handler,
    		// so it's only run when the user triggers this
    		selectedItems.add(item);
    	};
    }
    ```

    ---

    ```js hl_lines="4"
    function Dropdown() {
    	const selectedItems = new Set();
    	useEffect(() => {
    		// this code is inside of an Effect, so it only runs after rendering
    		logForAnalytics(selectedItems);
    	}, [selectedItems]);
    }
    ```

## –ö–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –∏ —Ö—É–∫–∏ –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –∏–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω—ã–º–∏ {#components-and-hooks-must-be-idempotent}

–ö–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –¥–æ–ª–∂–Ω—ã –≤—Å–µ–≥–¥–∞ –≤–æ–∑–≤—Ä–∞—â–∞—Ç—å –æ–¥–∏–Ω –∏ —Ç–æ—Ç –∂–µ —Ä–µ–∑—É–ª—å—Ç–∞—Ç –ø–æ –æ—Ç–Ω–æ—à–µ–Ω–∏—é –∫ —Å–≤–æ–∏–º –≤—Ö–æ–¥–Ω—ã–º –¥–∞–Ω–Ω—ã–º - —Ä–µ–∫–≤–∏–∑–∏—Ç–∞–º, —Å–æ—Å—Ç–æ—è–Ω–∏—é –∏ –∫–æ–Ω—Ç–µ–∫—Å—Ç—É. –≠—Ç–æ –∏–∑–≤–µ—Å—Ç–Ω–æ –∫–∞–∫ _–∏–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω–æ—Å—Ç—å_. [–ò–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω–æ—Å—Ç—å](https://ru.wikipedia.org/wiki/–ò–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω–æ—Å—Ç—å) - —Ç–µ—Ä–º–∏–Ω, –ø–æ–ª—É—á–∏–≤—à–∏–π —Ä–∞—Å–ø—Ä–æ—Å—Ç—Ä–∞–Ω–µ–Ω–∏–µ –≤ —Ñ—É–Ω–∫—Ü–∏–æ–Ω–∞–ª—å–Ω–æ–º –ø—Ä–æ–≥—Ä–∞–º–º–∏—Ä–æ–≤–∞–Ω–∏–∏. –û–Ω –æ—Ç–Ω–æ—Å–∏—Ç—Å—è –∫ –∏–¥–µ–µ, —á—Ç–æ –≤—ã [–≤—Å–µ–≥–¥–∞ –ø–æ–ª—É—á–∞–µ—Ç–µ –æ–¥–∏–Ω –∏ —Ç–æ—Ç –∂–µ —Ä–µ–∑—É–ª—å—Ç–∞—Ç –∫–∞–∂–¥—ã–π —Ä–∞–∑](../../learn/keeping-components-pure.md), –∫–æ–≥–¥–∞ –∑–∞–ø—É—Å–∫–∞–µ—Ç–µ —ç—Ç–æ—Ç –∫—É—Å–æ–∫ –∫–æ–¥–∞ —Å –æ–¥–Ω–∏–º–∏ –∏ —Ç–µ–º–∏ –∂–µ –≤—Ö–æ–¥–Ω—ã–º–∏ –¥–∞–Ω–Ω—ã–º–∏.

–≠—Ç–æ –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ _–≤–µ—Å—å_ –∫–æ–¥, –∫–æ—Ç–æ—Ä—ã–π –≤—ã–ø–æ–ª–Ω—è–µ—Ç—Å—è [–≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞](#how-does-react-run-your-code), –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å –∏–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω—ã–º, —á—Ç–æ–±—ã —ç—Ç–æ –ø—Ä–∞–≤–∏–ª–æ –≤—ã–ø–æ–ª–Ω—è–ª–æ—Å—å. –ù–∞–ø—Ä–∏–º–µ—Ä, —ç—Ç–∞ —Å—Ç—Ä–æ–∫–∞ –∫–æ–¥–∞ –Ω–µ —è–≤–ª—è–µ—Ç—Å—è –∏–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω–æ–π (–∏, —Å–ª–µ–¥–æ–≤–∞—Ç–µ–ª—å–Ω–æ, –∫–æ–º–ø–æ–Ω–µ–Ω—Ç —Ç–æ–∂–µ):

```js hl_lines="2"
function Clock() {
    const time = new Date(); // üî¥ Bad: always returns a different result!
    return <span>{time.toLocaleString()}</span>;
}
```

–§—É–Ω–∫—Ü–∏—è `new Date()` –Ω–µ —è–≤–ª—è–µ—Ç—Å—è –∏–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω–æ–π, –ø–æ—Å–∫–æ–ª—å–∫—É –≤—Å–µ–≥–¥–∞ –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç —Ç–µ–∫—É—â—É—é –¥–∞—Ç—É –∏ –º–µ–Ω—è–µ—Ç —Å–≤–æ–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç –ø—Ä–∏ –∫–∞–∂–¥–æ–º –≤—ã–∑–æ–≤–µ. –ü—Ä–∏ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–µ –≤—ã—à–µ—É–∫–∞–∑–∞–Ω–Ω–æ–≥–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ –≤—Ä–µ–º—è, –æ—Ç–æ–±—Ä–∞–∂–∞–µ–º–æ–µ –Ω–∞ —ç–∫—Ä–∞–Ω–µ, –±—É–¥–µ—Ç –∑–∞—Ü–∏–∫–ª–µ–Ω–æ –Ω–∞ —Ç–æ–º –≤—Ä–µ–º–µ–Ω–∏, –∫–æ–≥–¥–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç –±—ã–ª —Ä–µ–Ω–¥–µ—Ä–∏—Ä–æ–≤–∞–Ω. –ê–Ω–∞–ª–æ–≥–∏—á–Ω–æ, —Ñ—É–Ω–∫—Ü–∏–∏ –≤—Ä–æ–¥–µ `Math.random()` —Ç–∞–∫–∂–µ –Ω–µ —è–≤–ª—è—é—Ç—Å—è –∏–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω—ã–º–∏, –ø–æ—Å–∫–æ–ª—å–∫—É –ø—Ä–∏ –∫–∞–∂–¥–æ–º –≤—ã–∑–æ–≤–µ –æ–Ω–∏ –≤–æ–∑–≤—Ä–∞—â–∞—é—Ç —Ä–∞–∑–Ω—ã–µ —Ä–µ–∑—É–ª—å—Ç–∞—Ç—ã, –¥–∞–∂–µ –µ—Å–ª–∏ –≤—Ö–æ–¥–Ω—ã–µ –¥–∞–Ω–Ω—ã–µ –æ–¥–∏–Ω–∞–∫–æ–≤—ã.

–≠—Ç–æ –Ω–µ –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ –≤—ã –Ω–µ –¥–æ–ª–∂–Ω—ã –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –Ω–µ–∏–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω—ã–µ —Ñ—É–Ω–∫—Ü–∏–∏ –≤—Ä–æ–¥–µ `new Date()` –≤–æ–æ–±—â–µ - –≤—ã –ø—Ä–æ—Å—Ç–æ –¥–æ–ª–∂–Ω—ã –∏–∑–±–µ–≥–∞—Ç—å –∏—Ö –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏—è [–≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞](#how-does-react-run-your-code). –í –¥–∞–Ω–Ω–æ–º —Å–ª—É—á–∞–µ –º—ã –º–æ–∂–µ–º _—Å–∏–Ω—Ö—Ä–æ–Ω–∏–∑–∏—Ä–æ–≤–∞—Ç—å_ –ø–æ—Å–ª–µ–¥–Ω—é—é –¥–∞—Ç—É –¥–ª—è —ç—Ç–æ–≥–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ —Å –ø–æ–º–æ—â—å—é [Effect](../react/useEffect.md):

=== "App.js"

    ```js
    import { useState, useEffect } from 'react';

    function useTime() {
    	// 1. Keep track of the current date's state. `useState`
    	//    receives an initializer function as its
    	//    initial state. It only runs once when the hook is called,
    	//    so only the current date at the
    	//    time the hook is called is set first.
    	const [time, setTime] = useState(() => new Date());

    	useEffect(() => {
    		// 2. Update the current date every second using `setInterval`.
    		const id = setInterval(() => {
    			setTime(new Date()); // ‚úÖ Good: non-idempotent code
    			// no longer runs in render
    		}, 1000);
    		// 3. Return a cleanup function so we don't leak the
    		// `setInterval` timer.
    		return () => clearInterval(id);
    	}, []);

    	return time;
    }

    export default function Clock() {
    	const time = useTime();
    	return <span>{time.toLocaleString()}</span>;
    }
    ```

=== "CodeSandbox"

    <iframe src="https://codesandbox.io/embed/wp6f4j?view=Editor+%2B+Preview&module=%2Fsrc%2FApp.js" style="width:100%; height: 500px; border:0; border-radius: 4px; overflow:hidden;" title="react.dev" allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking" sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"></iframe>

–û–±–µ—Ä–Ω—É–≤ –Ω–µ—ç–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω—ã–π –≤—ã–∑–æ–≤ `new Date()` –≤ Effect, –æ–Ω –ø–µ—Ä–µ–Ω–æ—Å–∏—Ç —ç—Ç–æ –≤—ã—á–∏—Å–ª–µ–Ω–∏–µ [–∑–∞ –ø—Ä–µ–¥–µ–ª—ã —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞](#how-does-react-run-your-code).

–ï—Å–ª–∏ –≤–∞–º –Ω–µ –Ω—É–∂–Ω–æ —Å–∏–Ω—Ö—Ä–æ–Ω–∏–∑–∏—Ä–æ–≤–∞—Ç—å –≤–Ω–µ—à–Ω–µ–µ —Å–æ—Å—Ç–æ—è–Ω–∏–µ —Å React, –≤—ã —Ç–∞–∫–∂–µ –º–æ–∂–µ—Ç–µ —Ä–∞—Å—Å–º–æ—Ç—Ä–µ—Ç—å –≤–æ–∑–º–æ–∂–Ω–æ—Å—Ç—å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏—è [–æ–±—Ä–∞–±–æ—Ç—á–∏–∫–∞ —Å–æ–±—ã—Ç–∏–π](../../learn/responding-to-events.md), –µ—Å–ª–∏ –æ–Ω–æ –¥–æ–ª–∂–Ω–æ –æ–±–Ω–æ–≤–ª—è—Ç—å—Å—è —Ç–æ–ª—å–∫–æ –≤ –æ—Ç–≤–µ—Ç –Ω–∞ –≤–∑–∞–∏–º–æ–¥–µ–π—Å—Ç–≤–∏–µ —Å –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–µ–º.

## –ü–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã –¥–æ–ª–∂–Ω—ã –≤—ã–ø–æ–ª–Ω—è—Ç—å—Å—è –≤–Ω–µ —Ä–µ–Ω–¥–µ—Ä–∞ {#side-effects-must-run-outside-of-render}

[–ü–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã](../../learn/keeping-components-pure.md#side-effects-unintended-consequences) –Ω–µ –¥–æ–ª–∂–Ω—ã –∑–∞–ø—É—Å–∫–∞—Ç—å—Å—è [–≤ —Ä–µ–Ω–¥–µ—Ä–µ](#how-does-react-run-your-code), —Ç–∞–∫ –∫–∞–∫ React –º–æ–∂–µ—Ç —Ä–µ–Ω–¥–µ—Ä–∏—Ç—å –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –Ω–µ—Å–∫–æ–ª—å–∫–æ —Ä–∞–∑, —á—Ç–æ–±—ã —Å–æ–∑–¥–∞—Ç—å –Ω–∞–∏–ª—É—á—à–∏–π –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–π –æ–ø—ã—Ç.

!!!note "–ü–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã"

    –ü–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã - —ç—Ç–æ –±–æ–ª–µ–µ —à–∏—Ä–æ–∫–æ–µ –ø–æ–Ω—è—Ç–∏–µ, —á–µ–º —ç—Ñ—Ñ–µ–∫—Ç—ã. –≠—Ñ—Ñ–µ–∫—Ç—ã –æ—Ç–Ω–æ—Å—è—Ç—Å—è –∫ –∫–æ–¥—É, –æ–±–µ—Ä–Ω—É—Ç–æ–º—É –≤ `useEffect`, –≤ —Ç–æ –≤—Ä–µ–º—è –∫–∞–∫ –ø–æ–±–æ—á–Ω—ã–π —ç—Ñ—Ñ–µ–∫—Ç - —ç—Ç–æ –æ–±—â–∏–π —Ç–µ—Ä–º–∏–Ω –¥–ª—è –∫–æ–¥–∞, –∫–æ—Ç–æ—Ä—ã–π –∏–º–µ–µ—Ç –ª—é–±–æ–π –Ω–∞–±–ª—é–¥–∞–µ–º—ã–π —ç—Ñ—Ñ–µ–∫—Ç, –∫—Ä–æ–º–µ –µ–≥–æ –æ—Å–Ω–æ–≤–Ω–æ–≥–æ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞ - –≤–æ–∑–≤—Ä–∞—Ç–∞ –∑–Ω–∞—á–µ–Ω–∏—è –≤—ã–∑—ã–≤–∞—é—â–µ–π —Å—Ç–æ—Ä–æ–Ω–µ.

    –ü–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã –æ–±—ã—á–Ω–æ –ø–∏—à—É—Ç—Å—è –≤–Ω—É—Ç—Ä–∏ [–æ–±—Ä–∞–±–æ—Ç—á–∏–∫–æ–≤ —Å–æ–±—ã—Ç–∏–π](../../learn/responding-to-events.md) –∏–ª–∏ –≠—Ñ—Ñ–µ–∫—Ç–æ–≤. –ù–æ –Ω–∏–∫–æ–≥–¥–∞ - –≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞.

–•–æ—Ç—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥ –¥–æ–ª–∂–µ–Ω –±—ã—Ç—å —á–∏—Å—Ç—ã–º, –ø–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã –Ω–µ–æ–±—Ö–æ–¥–∏–º—ã, —á—Ç–æ–±—ã –≤ –∫–∞–∫–æ–π-—Ç–æ –º–æ–º–µ–Ω—Ç –≤–∞—à–µ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏–µ —Å–¥–µ–ª–∞–ª–æ —á—Ç–æ-–Ω–∏–±—É–¥—å –∏–Ω—Ç–µ—Ä–µ—Å–Ω–æ–µ, –Ω–∞–ø—Ä–∏–º–µ—Ä –ø–æ–∫–∞–∑–∞–ª–æ —á—Ç–æ-—Ç–æ –Ω–∞ —ç–∫—Ä–∞–Ω–µ! –ö–ª—é—á–µ–≤–æ–π –º–æ–º–µ–Ω—Ç —ç—Ç–æ–≥–æ –ø—Ä–∞–≤–∏–ª–∞ –∑–∞–∫–ª—é—á–∞–µ—Ç—Å—è –≤ —Ç–æ–º, —á—Ç–æ –ø–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã –Ω–µ –¥–æ–ª–∂–Ω—ã –≤—ã–ø–æ–ª–Ω—è—Ç—å—Å—è [–≤ —Ä–µ–Ω–¥–µ—Ä–µ](#how-does-react-run-your-code), —Ç–∞–∫ –∫–∞–∫ React –º–æ–∂–µ—Ç —Ä–µ–Ω–¥–µ—Ä–∏—Ç—å –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –Ω–µ—Å–∫–æ–ª—å–∫–æ —Ä–∞–∑. –í –±–æ–ª—å—à–∏–Ω—Å—Ç–≤–µ —Å–ª—É—á–∞–µ–≤ –¥–ª—è –æ–±—Ä–∞–±–æ—Ç–∫–∏ –ø–æ–±–æ—á–Ω—ã—Ö —ç—Ñ—Ñ–µ–∫—Ç–æ–≤ –≤—ã –±—É–¥–µ—Ç–µ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å [–æ–±—Ä–∞–±–æ—Ç—á–∏–∫–∏ —Å–æ–±—ã—Ç–∏–π](../../learn/responding-to-events.md). –ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ –æ–±—Ä–∞–±–æ—Ç—á–∏–∫–∞ —Å–æ–±—ã—Ç–∏–π —è–≤–Ω–æ –≥–æ–≤–æ—Ä–∏—Ç React, —á—Ç–æ —ç—Ç–æ—Ç –∫–æ–¥ –Ω–µ –Ω—É–∂–Ω–æ –∑–∞–ø—É—Å–∫–∞—Ç—å –≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞, —Å–æ—Ö—Ä–∞–Ω—è—è —á–∏—Å—Ç–æ—Ç—É —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞. –ï—Å–ª–∏ –≤—ã –∏—Å—á–µ—Ä–ø–∞–ª–∏ –≤—Å–µ –≤–∞—Ä–∏–∞–Ω—Ç—ã - –∏ —Ç–æ–ª—å–∫–æ –≤ –∫—Ä–∞–π–Ω–µ–º —Å–ª—É—á–∞–µ - –≤—ã —Ç–∞–∫–∂–µ –º–æ–∂–µ—Ç–µ –æ–±—Ä–∞–±–∞—Ç—ã–≤–∞—Ç—å –ø–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã —Å –ø–æ–º–æ—â—å—é `useEffect`.

### –ö–æ–≥–¥–∞ –º–æ–∂–Ω–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –º—É—Ç–∞—Ü–∏—é? {#mutation}

#### –õ–æ–∫–∞–ª—å–Ω–∞—è –º—É—Ç–∞—Ü–∏—è {#local-mutation}

–û–¥–Ω–∏–º –∏–∑ —Ä–∞—Å–ø—Ä–æ—Å—Ç—Ä–∞–Ω–µ–Ω–Ω—ã—Ö –ø—Ä–∏–º–µ—Ä–æ–≤ –ø–æ–±–æ—á–Ω–æ–≥–æ —ç—Ñ—Ñ–µ–∫—Ç–∞ —è–≤–ª—è–µ—Ç—Å—è –º—É—Ç–∞—Ü–∏—è, –∫–æ—Ç–æ—Ä–∞—è –≤ JavaScript –æ–∑–Ω–∞—á–∞–µ—Ç –∏–∑–º–µ–Ω–µ–Ω–∏–µ –∑–Ω–∞—á–µ–Ω–∏—è –Ω–µ [–ø—Ä–∏–º–∏—Ç–∏–≤–Ω–æ–≥–æ](https://developer.mozilla.org/docs/Glossary/Primitive) –∑–Ω–∞—á–µ–Ω–∏—è. –í —Ü–µ–ª–æ–º, —Ö–æ—Ç—è –º—É—Ç–∞—Ü–∏—è –Ω–µ —è–≤–ª—è–µ—Ç—Å—è –∏–¥–∏–æ–º–∞—Ç–∏—á–µ—Å–∫–æ–π –≤ React, _–ª–æ–∫–∞–ª—å–Ω–∞—è_ –º—É—Ç–∞—Ü–∏—è –∞–±—Å–æ–ª—é—Ç–Ω–æ –Ω–æ—Ä–º–∞–ª—å–Ω–∞:

```js hl_lines="2 7"
function FriendList({ friends }) {
    const items = []; // ‚úÖ Good: locally created
    for (let i = 0; i < friends.length; i++) {
        const friend = friends[i];
        items.push(
            <Friend key={friend.id} friend={friend} />
        ); // ‚úÖ Good: local mutation is okay
    }
    return <section>{items}</section>;
}
```

–ù–µ—Ç –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ—Å—Ç–∏ –∏—Å–∫–∞–∂–∞—Ç—å –∫–æ–¥, —á—Ç–æ–±—ã –∏–∑–±–µ–∂–∞—Ç—å –ª–æ–∫–∞–ª—å–Ω–æ–π –º—É—Ç–∞—Ü–∏–∏. –î–ª—è –∫—Ä–∞—Ç–∫–æ—Å—Ç–∏ –∑–¥–µ—Å—å –º–æ–∂–Ω–æ –±—ã–ª–æ –±—ã –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –∏ [`Array.map`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/map), –Ω–æ –Ω–µ—Ç –Ω–∏—á–µ–≥–æ –ø–ª–æ—Ö–æ–≥–æ –≤ —Ç–æ–º, —á—Ç–æ–±—ã —Å–æ–∑–¥–∞—Ç—å –ª–æ–∫–∞–ª—å–Ω—ã–π –º–∞—Å—Å–∏–≤ –∏ –∑–∞—Ç–µ–º –∑–∞—Ç–∞–ª–∫–∏–≤–∞—Ç—å –≤ –Ω–µ–≥–æ —ç–ª–µ–º–µ–Ω—Ç—ã [–≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞](#how-does-react-run-your-code).

–ù–µ—Å–º–æ—Ç—Ä—è –Ω–∞ —Ç–æ, —á—Ç–æ –≤—ã–≥–ª—è–¥–∏—Ç —Ç–∞–∫, –±—É–¥—Ç–æ –º—ã –º—É—Ç–∏—Ä—É–µ–º `items`, –≤–∞–∂–Ω–æ –æ—Ç–º–µ—Ç–∏—Ç—å, —á—Ç–æ —ç—Ç–æ—Ç –∫–æ–¥ –¥–µ–ª–∞–µ—Ç —ç—Ç–æ —Ç–æ–ª—å–∫–æ _–ª–æ–∫–∞–ª—å–Ω–æ_ - –º—É—Ç–∞—Ü–∏—è –Ω–µ ¬´–∑–∞–ø–æ–º–∏–Ω–∞–µ—Ç—Å—è¬ª –ø—Ä–∏ –ø–æ–≤—Ç–æ—Ä–Ω–æ–º —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞. –î—Ä—É–≥–∏–º–∏ —Å–ª–æ–≤–∞–º–∏, `items` –±—É–¥–µ—Ç —Å—É—â–µ—Å—Ç–≤–æ–≤–∞—Ç—å —Ç–æ–ª—å–∫–æ –¥–æ —Ç–µ—Ö –ø–æ—Ä, –ø–æ–∫–∞ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç –∫–æ–º–ø–æ–Ω–µ–Ω—Ç. –ü–æ—Å–∫–æ–ª—å–∫—É `items` –≤—Å–µ–≥–¥–∞ _–≤–æ—Å—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ—Ç—Å—è_ –ø—Ä–∏ –∫–∞–∂–¥–æ–º —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–µ `<FriendList />`, –∫–æ–º–ø–æ–Ω–µ–Ω—Ç –≤—Å–µ–≥–¥–∞ –±—É–¥–µ—Ç –≤–æ–∑–≤—Ä–∞—â–∞—Ç—å –æ–¥–∏–Ω –∏ —Ç–æ—Ç –∂–µ —Ä–µ–∑—É–ª—å—Ç–∞—Ç.

–° –¥—Ä—É–≥–æ–π —Å—Ç–æ—Ä–æ–Ω—ã, –µ—Å–ª–∏ `items` –±—ã–ª —Å–æ–∑–¥–∞–Ω –≤–Ω–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞, –æ–Ω —Å–æ—Ö—Ä–∞–Ω—è–µ—Ç —Å–≤–æ–∏ –ø—Ä–µ–¥—ã–¥—É—â–∏–µ –∑–Ω–∞—á–µ–Ω–∏—è –∏ –∑–∞–ø–æ–º–∏–Ω–∞–µ—Ç –∏–∑–º–µ–Ω–µ–Ω–∏—è:

```js hl_lines="1 7"
const items = []; // üî¥ Bad: created outside of the component
function FriendList({ friends }) {
    for (let i = 0; i < friends.length; i++) {
        const friend = friends[i];
        items.push(
            <Friend key={friend.id} friend={friend} />
        ); // üî¥ Bad: mutates a value created outside of render
    }
    return <section>{items}</section>;
}
```

–ö–æ–≥–¥–∞ `<FriendList />` –±—É–¥–µ—Ç –∑–∞–ø—É—â–µ–Ω —Å–Ω–æ–≤–∞, –º—ã –ø—Ä–æ–¥–æ–ª–∂–∏–º –¥–æ–±–∞–≤–ª—è—Ç—å `friends` –∫ `items` –ø—Ä–∏ –∫–∞–∂–¥–æ–º –∑–∞–ø—É—Å–∫–µ —ç—Ç–æ–≥–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞, —á—Ç–æ –ø—Ä–∏–≤–µ–¥–µ—Ç –∫ –º–Ω–æ–≥–æ—á–∏—Å–ª–µ–Ω–Ω—ã–º –¥—É–±–ª–∏—Ä—É—é—â–∏–º—Å—è —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞–º. –≠—Ç–∞ –≤–µ—Ä—Å–∏—è `<FriendList />` –∏–º–µ–µ—Ç –Ω–∞–±–ª—é–¥–∞–µ–º—ã–µ –ø–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã [–≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞](#how-does-react-run-your-code) –∏ **–Ω–∞—Ä—É—à–∞–µ—Ç –ø—Ä–∞–≤–∏–ª–æ**.

#### –õ–µ–Ω–∏–≤–∞—è –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è {#lazy-initialization}

–õ–µ–Ω–∏–≤–∞—è –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è —Ç–∞–∫–∂–µ –ø–æ–¥—Ö–æ–¥–∏—Ç, –Ω–µ—Å–º–æ—Ç—Ä—è –Ω–∞ —Ç–æ, —á—Ç–æ –Ω–µ —è–≤–ª—è–µ—Ç—Å—è –ø–æ–ª–Ω–æ—Å—Ç—å—é ¬´—á–∏—Å—Ç–æ–π¬ª:

```js hl_lines="2"
function ExpenseForm() {
    SuperCalculator.initializeIfNotReady(); // ‚úÖ Good: if it
    // doesn't affect other components
    // Continue rendering...
}
```

#### –ò–∑–º–µ–Ω–µ–Ω–∏–µ DOM {#changing-the-dom}

–í –ª–æ–≥–∏–∫–µ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ React –Ω–µ –¥–æ–ø—É—Å–∫–∞—é—Ç—Å—è –ø–æ–±–æ—á–Ω—ã–µ —ç—Ñ—Ñ–µ–∫—Ç—ã, –∫–æ—Ç–æ—Ä—ã–µ –≤–∏–¥–Ω—ã –Ω–µ–ø–æ—Å—Ä–µ–¥—Å—Ç–≤–µ–Ω–Ω–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—é. –î—Ä—É–≥–∏–º–∏ —Å–ª–æ–≤–∞–º–∏, –ø—Ä–æ—Å—Ç–æ–π –≤—ã–∑–æ–≤ —Ñ—É–Ω–∫—Ü–∏–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ —Å–∞–º –ø–æ —Å–µ–±–µ –Ω–µ –¥–æ–ª–∂–µ–Ω –ø—Ä–∏–≤–æ–¥–∏—Ç—å –∫ –∏–∑–º–µ–Ω–µ–Ω–∏—è–º –Ω–∞ —ç–∫—Ä–∞–Ω–µ.

```js hl_lines="2"
function ProductDetailPage({ product }) {
    document.window.title = product.title; // üî¥ Bad: Changes the DOM
}
```

–û–¥–∏–Ω –∏–∑ —Å–ø–æ—Å–æ–±–æ–≤ –¥–æ–±–∏—Ç—å—Å—è –∂–µ–ª–∞–µ–º–æ–≥–æ —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞ –æ–±–Ω–æ–≤–ª–µ–Ω–∏—è `window.title` –≤–Ω–µ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ - —ç—Ç–æ [—Å–∏–Ω—Ö—Ä–æ–Ω–∏–∑–∏—Ä–æ–≤–∞—Ç—å –∫–æ–º–ø–æ–Ω–µ–Ω—Ç —Å `window`](../../learn/synchronizing-with-effects.md).

–ü–æ–∫–∞ –≤—ã–∑–æ–≤ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ –Ω–µ—Å–∫–æ–ª—å–∫–æ —Ä–∞–∑ –±–µ–∑–æ–ø–∞—Å–µ–Ω –∏ –Ω–µ –≤–ª–∏—è–µ—Ç –Ω–∞ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥ –¥—Ä—É–≥–∏—Ö –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤, React –Ω–µ –≤–æ–ª–Ω—É–µ—Ç, —è–≤–ª—è–µ—Ç—Å—è –ª–∏ –æ–Ω –Ω–∞ 100% —á–∏—Å—Ç—ã–º –≤ —Å—Ç—Ä–æ–≥–æ–º —Å–º—ã—Å–ª–µ —Ñ—É–Ω–∫—Ü–∏–æ–Ω–∞–ª—å–Ω–æ–≥–æ –ø—Ä–æ–≥—Ä–∞–º–º–∏—Ä–æ–≤–∞–Ω–∏—è. –í–∞–∂–Ω–µ–µ, —á—Ç–æ [–∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –∏–¥–µ–º–ø–æ—Ç–µ–Ω—Ç–Ω—ã–º–∏](./components-and-hooks-must-be-pure.md).

## –ü—Ä–æ–ø—Å—ã –∏ —Å–æ—Å—Ç–æ—è–Ω–∏–µ –Ω–µ–∏–∑–º–µ–Ω—è–µ–º—ã {#props-and-state-are-immutable}

–ü—Ä–æ–ø—Å—ã –∏ —Å–æ—Å—Ç–æ—è–Ω–∏–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ –Ω–µ–∏–∑–º–µ–Ω—è–µ–º—ã [snapshots](../../learn/state-as-a-snapshot.md). –ù–∏–∫–æ–≥–¥–∞ –Ω–µ –∏–∑–º–µ–Ω—è–π—Ç–µ –∏—Ö –Ω–∞–ø—Ä—è–º—É—é. –í–º–µ—Å—Ç–æ —ç—Ç–æ–≥–æ –ø–µ—Ä–µ–¥–∞–≤–∞–π—Ç–µ –Ω–æ–≤—ã–µ –ø—Ä–æ–ø—Å—ã –≤–Ω–∏–∑ –∏ –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ —Ñ—É–Ω–∫—Ü–∏—é setter –∏–∑ `useState`.

–ú–æ–∂–Ω–æ —Ä–∞—Å—Å–º–∞—Ç—Ä–∏–≤–∞—Ç—å –ø—Ä–æ–ø—Å—ã –∏ –∑–Ω–∞—á–µ–Ω–∏—è —Å–æ—Å—Ç–æ—è–Ω–∏—è –∫–∞–∫ –º–æ–º–µ–Ω—Ç–∞–ª—å–Ω—ã–µ —Å–Ω–∏–º–∫–∏, –∫–æ—Ç–æ—Ä—ã–µ –æ–±–Ω–æ–≤–ª—è—é—Ç—Å—è –ø–æ—Å–ª–µ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞. –ü–æ —ç—Ç–æ–π –ø—Ä–∏—á–∏–Ω–µ –≤—ã –Ω–µ –∏–∑–º–µ–Ω—è–µ—Ç–µ –ø–µ—Ä–µ–º–µ–Ω–Ω—ã–µ props –∏–ª–∏ state –Ω–∞–ø—Ä—è–º—É—é: –≤–º–µ—Å—Ç–æ —ç—Ç–æ–≥–æ –≤—ã –ø–µ—Ä–µ–¥–∞–µ—Ç–µ –Ω–æ–≤—ã–µ props –∏–ª–∏ –∏—Å–ø–æ–ª—å–∑—É–µ—Ç–µ –ø—Ä–µ–¥–æ—Å—Ç–∞–≤–ª–µ–Ω–Ω—É—é –≤–∞–º —Ñ—É–Ω–∫—Ü–∏—é setter, —á—Ç–æ–±—ã —Å–æ–æ–±—â–∏—Ç—å React, —á—Ç–æ —Å–æ—Å—Ç–æ—è–Ω–∏–µ –¥–æ–ª–∂–Ω–æ –±—ã—Ç—å –æ–±–Ω–æ–≤–ª–µ–Ω–æ –ø—Ä–∏ —Å–ª–µ–¥—É—é—â–µ–º —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞.

### –ù–µ –º—É—Ç–∏—Ä—É–π—Ç–µ –ø—Ä–æ–ø—Å—ã {#props}

–ü—Ä–æ–ø—Å—ã –Ω–µ–∏–∑–º–µ–Ω—è–µ–º—ã, –ø–æ—Ç–æ–º—É —á—Ç–æ –µ—Å–ª–∏ –≤—ã –∏—Ö –∏–∑–º–µ–Ω–∏—Ç–µ, –ø—Ä–∏–ª–æ–∂–µ–Ω–∏–µ –±—É–¥–µ—Ç –≤—ã–¥–∞–≤–∞—Ç—å –Ω–µ–ø–æ—Å–ª–µ–¥–æ–≤–∞—Ç–µ–ª—å–Ω—ã–π –≤—ã–≤–æ–¥, –∫–æ—Ç–æ—Ä—ã–π –º–æ–∂–µ—Ç –±—ã—Ç—å —Ç—Ä—É–¥–Ω–æ –æ—Ç–ª–∞–¥–∏—Ç—å, –ø–æ—Å–∫–æ–ª—å–∫—É –æ–Ω –º–æ–∂–µ—Ç —Ä–∞–±–æ—Ç–∞—Ç—å –∏–ª–∏ –Ω–µ —Ä–∞–±–æ—Ç–∞—Ç—å –≤ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –æ—Ç –æ–±—Å—Ç–æ—è—Ç–µ–ª—å—Å—Ç–≤.

```js hl_lines="2"
function Post({ item }) {
    item.url = new Url(item.url, base); // üî¥ Bad: never mutate props directly
    return <Link url={item.url}>{item.title}</Link>;
}
```

---

```js hl_lines="2"
function Post({ item }) {
    const url = new Url(item.url, base); // ‚úÖ Good: make a copy instead
    return <Link url={url}>{item.title}</Link>;
}
```

### –ù–µ –º—É—Ç–∏—Ä—É–π—Ç–µ —Å–æ—Å—Ç–æ—è–Ω–∏–µ {#state}

`useState` –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç –ø–µ—Ä–µ–º–µ–Ω–Ω—É—é state –∏ —Å–µ—Ç—Ç–µ—Ä –¥–ª—è –æ–±–Ω–æ–≤–ª–µ–Ω–∏—è —ç—Ç–æ–≥–æ —Å–æ—Å—Ç–æ—è–Ω–∏—è.

```js
const [stateVariable, setter] = useState(0);
```

–í–º–µ—Å—Ç–æ —Ç–æ–≥–æ —á—Ç–æ–±—ã –æ–±–Ω–æ–≤–ª—è—Ç—å –ø–µ—Ä–µ–º–µ–Ω–Ω—É—é state –Ω–∞ –º–µ—Å—Ç–µ, –Ω–∞–º –Ω—É–∂–Ω–æ –æ–±–Ω–æ–≤–∏—Ç—å –µ–µ —Å –ø–æ–º–æ—â—å—é —Ñ—É–Ω–∫—Ü–∏–∏ setter, –∫–æ—Ç–æ—Ä—É—é –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç `useState`. –ò–∑–º–µ–Ω–µ–Ω–∏–µ –∑–Ω–∞—á–µ–Ω–∏—è –ø–µ—Ä–µ–º–µ–Ω–Ω–æ–π state –Ω–µ –ø—Ä–∏–≤–æ–¥–∏—Ç –∫ –æ–±–Ω–æ–≤–ª–µ–Ω–∏—é –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞, –æ—Å—Ç–∞–≤–ª—è—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª–µ–π —Å —É—Å—Ç–∞—Ä–µ–≤—à–∏–º –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–º –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–æ–º. –ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ —Ñ—É–Ω–∫—Ü–∏–∏ setter –∏–Ω—Ñ–æ—Ä–º–∏—Ä—É–µ—Ç React –æ —Ç–æ–º, —á—Ç–æ —Å–æ—Å—Ç–æ—è–Ω–∏–µ –∏–∑–º–µ–Ω–∏–ª–æ—Å—å, –∏ —á—Ç–æ –Ω–∞–º –Ω—É–∂–Ω–æ –ø–æ—Å—Ç–∞–≤–∏—Ç—å –≤ –æ—á–µ—Ä–µ–¥—å –ø–æ–≤—Ç–æ—Ä–Ω—ã–π —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥ –¥–ª—è –æ–±–Ω–æ–≤–ª–µ–Ω–∏—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞.

```js hl_lines="5"
function Counter() {
    const [count, setCount] = useState(0);

    function handleClick() {
        count = count + 1; // üî¥ Bad: never mutate state directly
    }

    return (
        <button onClick={handleClick}>
            You pressed me {count} times
        </button>
    );
}
```

---

```js hl_lines="5-6"
function Counter() {
    const [count, setCount] = useState(0);

    function handleClick() {
        setCount(count + 1); // ‚úÖ Good: use the setter function
        // returned by useState
    }

    return (
        <button onClick={handleClick}>
            You pressed me {count} times
        </button>
    );
}
```

## –í–æ–∑–≤—Ä–∞—â–∞–µ–º—ã–µ –∑–Ω–∞—á–µ–Ω–∏—è –∏ –∞—Ä–≥—É–º–µ–Ω—Ç—ã —Ö—É–∫–æ–≤ –Ω–µ–∏–∑–º–µ–Ω—è–µ–º—ã {#return-values-and-arguments-to-hooks-are-immutable}

–ü–æ—Å–ª–µ –ø–µ—Ä–µ–¥–∞—á–∏ –∑–Ω–∞—á–µ–Ω–∏–π –≤ —Ö—É–∫ –∏—Ö –Ω–µ–ª—å–∑—è –∏–∑–º–µ–Ω—è—Ç—å. –ö–∞–∫ –∏ –ø—Ä–æ–ø—Å—ã –≤ JSX, –∑–Ω–∞—á–µ–Ω–∏—è —Å—Ç–∞–Ω–æ–≤—è—Ç—Å—è –Ω–µ–∏–∑–º–µ–Ω—è–µ–º—ã–º–∏ –ø—Ä–∏ –ø–µ—Ä–µ–¥–∞—á–µ —Ö—É–∫—É.

```js hl_lines="4-5"
function useIconStyle(icon) {
    const theme = useContext(ThemeContext);
    if (icon.enabled) {
        // üî¥ Bad: never mutate hook arguments directly
        icon.className = computeStyle(icon, theme);
    }
    return icon;
}
```

---

```js hl_lines="3"
function useIconStyle(icon) {
    const theme = useContext(ThemeContext);
    const newIcon = { ...icon }; // ‚úÖ Good: make a copy instead
    if (icon.enabled) {
        newIcon.className = computeStyle(icon, theme);
    }
    return newIcon;
}
```

–û–¥–Ω–∏–º –∏–∑ –≤–∞–∂–Ω—ã—Ö –ø—Ä–∏–Ω—Ü–∏–ø–æ–≤ React —è–≤–ª—è–µ—Ç—Å—è _–ª–æ–∫–∞–ª—å–Ω–æ–µ —Ä–∞—Å—Å—É–∂–¥–µ–Ω–∏–µ_: —Å–ø–æ—Å–æ–±–Ω–æ—Å—Ç—å –ø–æ–Ω—è—Ç—å, —á—Ç–æ –¥–µ–ª–∞–µ—Ç –∫–æ–º–ø–æ–Ω–µ–Ω—Ç –∏–ª–∏ —Ö—É–∫, –≥–ª—è–¥—è –Ω–∞ –µ–≥–æ –∫–æ–¥ –≤ –∏–∑–æ–ª—è—Ü–∏–∏. –ö —Ö—É–∫–∞–º —Å–ª–µ–¥—É–µ—Ç –æ—Ç–Ω–æ—Å–∏—Ç—å—Å—è –∫–∞–∫ –∫ ¬´—á–µ—Ä–Ω—ã–º —è—â–∏–∫–∞–º¬ª, –∫–æ–≥–¥–∞ –æ–Ω–∏ –≤—ã–∑—ã–≤–∞—é—Ç—Å—è. –ù–∞–ø—Ä–∏–º–µ—Ä, –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–π —Ö—É–∫ –º–æ–∂–µ—Ç –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å —Å–≤–æ–∏ –∞—Ä–≥—É–º–µ–Ω—Ç—ã –≤ –∫–∞—á–µ—Å—Ç–≤–µ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π –¥–ª—è –º–µ–º–æ–∏–∑–∞—Ü–∏–∏ –∑–Ω–∞—á–µ–Ω–∏–π –≤–Ω—É—Ç—Ä–∏ –Ω–µ–≥–æ:

```js
function useIconStyle(icon) {
    const theme = useContext(ThemeContext);

    return useMemo(() => {
        const newIcon = { ...icon };
        if (icon.enabled) {
            newIcon.className = computeStyle(icon, theme);
        }
        return newIcon;
    }, [icon, theme]);
}
```

–ï—Å–ª–∏ –≤—ã –∏–∑–º–µ–Ω–∏—Ç–µ –∞—Ä–≥—É–º–µ–Ω—Ç—ã Hooks, –º–µ–º–æ–∏–∑–∞—Ü–∏—è –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ —Ö—É–∫–∞ —Å—Ç–∞–Ω–µ—Ç –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω–æ–π, –ø–æ—ç—Ç–æ–º—É –≤–∞–∂–Ω–æ –∏–∑–±–µ–≥–∞—Ç—å —ç—Ç–æ–≥–æ.

```js hl_lines="2"
style = useIconStyle(icon); // `style` is memoized based on `icon`
icon.enabled = false; // Bad: üî¥ never mutate hook arguments directly
style = useIconStyle(icon); // previously memoized result is returned
```

---

```js hl_lines="2"
style = useIconStyle(icon); // `style` is memoized based on `icon`
icon = { ...icon, enabled: false }; // Good: ‚úÖ make a copy instead
style = useIconStyle(icon); // new value of `style` is calculated
```

–ê–Ω–∞–ª–æ–≥–∏—á–Ω–æ, –≤–∞–∂–Ω–æ –Ω–µ –∏–∑–º–µ–Ω—è—Ç—å –≤–æ–∑–≤—Ä–∞—â–∞–µ–º—ã–µ –∑–Ω–∞—á–µ–Ω–∏—è —Ö—É–∫–æ–≤, —Ç–∞–∫ –∫–∞–∫ –æ–Ω–∏ –º–æ–≥—É—Ç –±—ã—Ç—å –º–µ–º–æ–∏–∑–∏—Ä–æ–≤–∞–Ω—ã.

## –ó–Ω–∞—á–µ–Ω–∏—è –Ω–µ–∏–∑–º–µ–Ω—è–µ–º—ã –ø–æ—Å–ª–µ –ø–µ—Ä–µ–¥–∞—á–∏ –≤ JSX {#values-are-immutable-after-being-passed-to-jsx}

–ù–µ –º—É—Ç–∏—Ä—É–π—Ç–µ –∑–Ω–∞—á–µ–Ω–∏—è –ø–æ—Å–ª–µ —Ç–æ–≥–æ, –∫–∞–∫ –æ–Ω–∏ –±—ã–ª–∏ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω—ã –≤ JSX. –ü–µ—Ä–µ–º–µ—Å—Ç–∏—Ç–µ –º—É—Ç–∞—Ü–∏—é –¥–æ —Å–æ–∑–¥–∞–Ω–∏—è JSX.

–ö–æ–≥–¥–∞ –≤—ã –∏—Å–ø–æ–ª—å–∑—É–µ—Ç–µ JSX –≤ –≤—ã—Ä–∞–∂–µ–Ω–∏–∏, React –º–æ–∂–µ—Ç –Ω–µ—Ç–µ—Ä–ø–µ–ª–∏–≤–æ –æ—Ü–µ–Ω–∏—Ç—å JSX –¥–æ —Ç–æ–≥–æ, –∫–∞–∫ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç –∑–∞–∫–æ–Ω—á–∏—Ç —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥. –≠—Ç–æ –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ –º—É—Ç–∞—Ü–∏—è –∑–Ω–∞—á–µ–Ω–∏–π –ø–æ—Å–ª–µ –∏—Ö –ø–µ—Ä–µ–¥–∞—á–∏ –≤ JSX –º–æ–∂–µ—Ç –ø—Ä–∏–≤–µ—Å—Ç–∏ –∫ —É—Å—Ç–∞—Ä–µ–≤—à–µ–º—É –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–º—É –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å—É, –ø–æ—Å–∫–æ–ª—å–∫—É React –Ω–µ –±—É–¥–µ—Ç –∑–Ω–∞—Ç—å, —á—Ç–æ –Ω—É–∂–Ω–æ –æ–±–Ω–æ–≤–∏—Ç—å –≤—ã–≤–æ–¥ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞.

```js hl_lines="4-5"
function Page({ colour }) {
    const styles = { colour, size: 'large' };
    const header = <Header styles={styles} />;
    // üî¥ Bad: styles was already used in the JSX above
    styles.size = 'small';
    const footer = <Footer styles={styles} />;
    return (
        <>
            {header}
            <Content />
            {footer}
        </>
    );
}
```

---

```js hl_lines="4-5"
function Page({ colour }) {
    const headerStyles = { colour, size: 'large' };
    const header = <Header styles={headerStyles} />;
    // ‚úÖ Good: we created a new value
    const footerStyles = { colour, size: 'small' };
    const footer = <Footer styles={footerStyles} />;
    return (
        <>
            {header}
            <Content />
            {footer}
        </>
    );
}
```

<small>:material-information-outline: –ò—Å—Ç–æ—á–Ω–∏–∫ &mdash; <https://react.dev/reference/rules/components-and-hooks-must-be-pure></small>
