---
description: React –æ—Ç–≤–µ—á–∞–µ—Ç –∑–∞ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ –∏ —Ö—É–∫–æ–≤, –∫–æ–≥–¥–∞ —ç—Ç–æ –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ –¥–ª—è –æ–ø—Ç–∏–º–∏–∑–∞—Ü–∏–∏ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ –æ–ø—ã—Ç–∞. –û–Ω —è–≤–ª—è–µ—Ç—Å—è –¥–µ–∫–ª–∞—Ä–∞—Ç–∏–≤–Ω—ã–º - –≤—ã —É–∫–∞–∑—ã–≤–∞–µ—Ç–µ React, —á—Ç–æ –Ω—É–∂–Ω–æ –æ—Ç–æ–±—Ä–∞–∑–∏—Ç—å –≤ –ª–æ–≥–∏–∫–µ –≤–∞—à–µ–≥–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞, –∞ React —Å–∞–º —Ä–µ—à–∞–µ—Ç, –∫–∞–∫ –ª—É—á—à–µ –ø–æ–∫–∞–∑–∞—Ç—å —ç—Ç–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—é
---

# React –≤—ã–∑—ã–≤–∞–µ—Ç –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –∏ —Ö—É–∫–∏

<big>
React –æ—Ç–≤–µ—á–∞–µ—Ç –∑–∞ —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ –∏ —Ö—É–∫–æ–≤, –∫–æ–≥–¥–∞ —ç—Ç–æ –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ –¥–ª—è –æ–ø—Ç–∏–º–∏–∑–∞—Ü–∏–∏ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ –æ–ø—ã—Ç–∞. –û–Ω —è–≤–ª—è–µ—Ç—Å—è –¥–µ–∫–ª–∞—Ä–∞—Ç–∏–≤–Ω—ã–º: –≤—ã —É–∫–∞–∑—ã–≤–∞–µ—Ç–µ React, —á—Ç–æ –Ω—É–∂–Ω–æ –æ—Ç–æ–±—Ä–∞–∑–∏—Ç—å –≤ –ª–æ–≥–∏–∫–µ –≤–∞—à–µ–≥–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞, –∞ React —Å–∞–º —Ä–µ—à–∞–µ—Ç, –∫–∞–∫ –ª—É—á—à–µ –ø–æ–∫–∞–∑–∞—Ç—å —ç—Ç–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—é.
</big>

## –ù–∏–∫–æ–≥–¥–∞ –Ω–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ñ—É–Ω–∫—Ü–∏–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ –Ω–∞–ø—Ä—è–º—É—é {#never-call-component-functions-directly}

–ö–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –¥–æ–ª–∂–Ω—ã –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å—Å—è —Ç–æ–ª—å–∫–æ –≤ JSX. –ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ –∏—Ö –∫–∞–∫ –æ–±—ã—á–Ω—ã–µ —Ñ—É–Ω–∫—Ü–∏–∏. –ò—Ö –¥–æ–ª–∂–µ–Ω –≤—ã–∑—ã–≤–∞—Ç—å React.

React –¥–æ–ª–∂–µ–Ω —Ä–µ—à–∏—Ç—å, –∫–æ–≥–¥–∞ –±—É–¥–µ—Ç –≤—ã–∑–≤–∞–Ω–∞ —Ñ—É–Ω–∫—Ü–∏—è –≤–∞—à–µ–≥–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ [–≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞](components-and-hooks-must-be-pure.md#how-does-react-run-your-code). –í React –≤—ã –¥–µ–ª–∞–µ—Ç–µ —ç—Ç–æ —Å –ø–æ–º–æ—â—å—é JSX.

```js hl_lines="6"
function BlogPost() {
    return (
        <Layout>
            <Article />
        </Layout>
    ); // ‚úÖ Good: Only use components in JSX
}
```

---

```js hl_lines="2"
function BlogPost() {
    return <Layout>{Article()}</Layout>; // üî¥ Bad: Never call them directly
}
```

–ï—Å–ª–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç —Å–æ–¥–µ—Ä–∂–∏—Ç —Ö—É–∫–∏, –º–æ–∂–Ω–æ –ª–µ–≥–∫–æ –Ω–∞—Ä—É—à–∏—Ç—å [–ü—Ä–∞–≤–∏–ª–∞ —Ö—É–∫–æ–≤](./rules-of-hooks.md), –∫–æ–≥–¥–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –≤—ã–∑—ã–≤–∞—é—Ç—Å—è –Ω–∞–ø—Ä—è–º—É—é –≤ —Ü–∏–∫–ª–µ –∏–ª–∏ —É—Å–ª–æ–≤–Ω–æ.

–ü–æ–∑–≤–æ–ª—è—è React –æ—Ä–∫–µ—Å—Ç—Ä–æ–≤–∞—Ç—å —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥, –≤—ã —Ç–∞–∫–∂–µ –ø–æ–ª—É—á–∞–µ—Ç–µ —Ä—è–¥ –ø—Ä–µ–∏–º—É—â–µ—Å—Ç–≤:

-   **–ö–æ–º–ø–æ–Ω–µ–Ω—Ç—ã —Å—Ç–∞–Ω–æ–≤—è—Ç—Å—è –±–æ–ª—å—à–µ —á–µ–º —Ñ—É–Ω–∫—Ü–∏—è–º–∏.** React –º–æ–∂–µ—Ç –¥–æ–ø–æ–ª–Ω–∏—Ç—å –∏—Ö —Ç–∞–∫–∏–º–∏ –≤–æ–∑–º–æ–∂–Ω–æ—Å—Ç—è–º–∏, –∫–∞–∫ _–ª–æ–∫–∞–ª—å–Ω–æ–µ —Å–æ—Å—Ç–æ—è–Ω–∏–µ_ —Å –ø–æ–º–æ—â—å—é —Ö—É–∫–æ–≤, –∫–æ—Ç–æ—Ä—ã–µ –ø—Ä–∏–≤—è–∑–∞–Ω—ã –∫ –∏–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ç–æ—Ä—É –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ –≤ –¥–µ—Ä–µ–≤–µ.
-   **–¢–∏–ø—ã –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ —É—á–∞—Å—Ç–≤—É—é—Ç –≤ —Å–æ–≥–ª–∞—Å–æ–≤–∞–Ω–∏–∏.** –ü–æ–∑–≤–æ–ª—è—è React –≤—ã–∑—ã–≤–∞—Ç—å –≤–∞—à–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã, –≤—ã —Ç–∞–∫–∂–µ —Å–æ–æ–±—â–∞–µ—Ç–µ –µ–º—É –±–æ–ª—å—à–µ –æ –∫–æ–Ω—Ü–µ–ø—Ç—É–∞–ª—å–Ω–æ–π —Å—Ç—Ä—É–∫—Ç—É—Ä–µ –≤–∞—à–µ–≥–æ –¥–µ—Ä–µ–≤–∞. –ù–∞–ø—Ä–∏–º–µ—Ä, –∫–æ–≥–¥–∞ –≤—ã –ø–µ—Ä–µ—Ö–æ–¥–∏—Ç–µ –æ—Ç —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ `<Feed>` –∫ —Å—Ç—Ä–∞–Ω–∏—Ü–µ `<Profile>`, React –Ω–µ –±—É–¥–µ—Ç –ø—ã—Ç–∞—Ç—å—Å—è –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –∏—Ö –ø–æ–≤—Ç–æ—Ä–Ω–æ.
-   **React –º–æ–∂–µ—Ç —É–ª—É—á—à–∏—Ç—å –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–π –æ–ø—ã—Ç.** –ù–∞–ø—Ä–∏–º–µ—Ä, –æ–Ω –º–æ–∂–µ—Ç –ø–æ–∑–≤–æ–ª–∏—Ç—å –±—Ä–∞—É–∑–µ—Ä—É –≤—ã–ø–æ–ª–Ω—è—Ç—å –Ω–µ–∫–æ—Ç–æ—Ä—É—é —Ä–∞–±–æ—Ç—É –º–µ–∂–¥—É –≤—ã–∑–æ–≤–∞–º–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤, —á—Ç–æ–±—ã –ø–æ–≤—Ç–æ—Ä–Ω—ã–π —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥ –±–æ–ª—å—à–æ–≥–æ –¥–µ—Ä–µ–≤–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ –Ω–µ –±–ª–æ–∫–∏—Ä–æ–≤–∞–ª –æ—Å–Ω–æ–≤–Ω–æ–π –ø–æ—Ç–æ–∫.
-   **–õ—É—á—à–∞—è –∏—Å—Ç–æ—Ä–∏—è –æ—Ç–ª–∞–¥–∫–∏.** –ï—Å–ª–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã —è–≤–ª—è—é—Ç—Å—è –≥—Ä–∞–∂–¥–∞–Ω–∞–º–∏ –ø–µ—Ä–≤–æ–≥–æ –∫–ª–∞—Å—Å–∞, –æ –∫–æ—Ç–æ—Ä—ã—Ö –∑–Ω–∞–µ—Ç –±–∏–±–ª–∏–æ—Ç–µ–∫–∞, –º—ã –º–æ–∂–µ–º —Å–æ–∑–¥–∞—Ç—å –±–æ–≥–∞—Ç—ã–µ –∏–Ω—Å—Ç—Ä—É–º–µ–Ω—Ç—ã —Ä–∞–∑—Ä–∞–±–æ—Ç—á–∏–∫–∞ –¥–ª—è –∏–Ω—Ç—Ä–æ—Å–ø–µ–∫—Ü–∏–∏ –≤ –ø—Ä–æ—Ü–µ—Å—Å–µ —Ä–∞–∑—Ä–∞–±–æ—Ç–∫–∏.
-   **–ë–æ–ª–µ–µ —ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω–æ–µ —Å–æ–≥–ª–∞—Å–æ–≤–∞–Ω–∏–µ.** React –º–æ–∂–µ—Ç —Ç–æ—á–Ω–æ –æ–ø—Ä–µ–¥–µ–ª–∏—Ç—å, –∫–∞–∫–∏–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã –≤ –¥–µ—Ä–µ–≤–µ –Ω—É–∂–¥–∞—é—Ç—Å—è –≤ –ø–æ–≤—Ç–æ—Ä–Ω–æ–º —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–µ, –∏ –ø—Ä–æ–ø—É—Å—Ç–∏—Ç—å —Ç–µ, –∫–æ—Ç–æ—Ä—ã–µ –Ω–µ –Ω—É–∂–¥–∞—é—Ç—Å—è. –≠—Ç–æ –¥–µ–ª–∞–µ—Ç –≤–∞—à–µ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏–µ –±–æ–ª–µ–µ –±—ã—Å—Ç—Ä—ã–º –∏ —à—É—Å—Ç—Ä—ã–º.

## –ù–∏–∫–æ–≥–¥–∞ –Ω–µ –ø–µ—Ä–µ–¥–∞–≤–∞–π—Ç–µ —Ö—É–∫–∏ –∫–∞–∫ –æ–±—ã—á–Ω—ã–µ –∑–Ω–∞—á–µ–Ω–∏—è {#never-pass-around-hooks-as-regular-values}

–•—É–∫–∏ –¥–æ–ª–∂–Ω—ã –≤—ã–∑—ã–≤–∞—Ç—å—Å—è —Ç–æ–ª—å–∫–æ –≤–Ω—É—Ç—Ä–∏ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ –∏–ª–∏ —Ö—É–∫–æ–≤. –ù–∏–∫–æ–≥–¥–∞ –Ω–µ –ø–µ—Ä–µ–¥–∞–≤–∞–π—Ç–µ –∏—Ö –∫–∞–∫ –æ–±—ã—á–Ω—ã–µ –∑–Ω–∞—á–µ–Ω–∏—è.

–•—É–∫–∏ –ø–æ–∑–≤–æ–ª—è—é—Ç –¥–æ–ø–æ–ª–Ω–∏—Ç—å –∫–æ–º–ø–æ–Ω–µ–Ω—Ç —Ñ—É–Ω–∫—Ü–∏—è–º–∏ React. –û–Ω–∏ –≤—Å–µ–≥–¥–∞ –¥–æ–ª–∂–Ω—ã –≤—ã–∑—ã–≤–∞—Ç—å—Å—è –∫–∞–∫ —Ñ—É–Ω–∫—Ü–∏—è –∏ –Ω–∏–∫–æ–≥–¥–∞ –Ω–µ –ø–µ—Ä–µ–¥–∞–≤–∞—Ç—å—Å—è –∫–∞–∫ –æ–±—ã—á–Ω–æ–µ –∑–Ω–∞—á–µ–Ω–∏–µ. –≠—Ç–æ –ø–æ–∑–≤–æ–ª—è–µ—Ç –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å _–ª–æ–∫–∞–ª—å–Ω—ã–µ —Ä–∞—Å—Å—É–∂–¥–µ–Ω–∏—è_, –∏–ª–∏ –≤–æ–∑–º–æ–∂–Ω–æ—Å—Ç—å –¥–ª—è —Ä–∞–∑—Ä–∞–±–æ—Ç—á–∏–∫–æ–≤ –ø–æ–Ω—è—Ç—å –≤—Å–µ, —á—Ç–æ –º–æ–∂–µ—Ç –¥–µ–ª–∞—Ç—å –∫–æ–º–ø–æ–Ω–µ–Ω—Ç, —Ä–∞—Å—Å–º–∞—Ç—Ä–∏–≤–∞—è –µ–≥–æ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ—Å—Ç–∏.

–ù–∞—Ä—É—à–µ–Ω–∏–µ —ç—Ç–æ–≥–æ –ø—Ä–∞–≤–∏–ª–∞ –ø—Ä–∏–≤–µ–¥–µ—Ç –∫ —Ç–æ–º—É, —á—Ç–æ React –Ω–µ –±—É–¥–µ—Ç –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏ –æ–ø—Ç–∏–º–∏–∑–∏—Ä–æ–≤–∞—Ç—å –≤–∞—à –∫–æ–º–ø–æ–Ω–µ–Ω—Ç.

### –ù–µ –º—É—Ç–∏—Ä—É–π—Ç–µ –¥–∏–Ω–∞–º–∏—á–µ—Å–∫–∏ —Ö—É–∫–∏ {#dont-dynamically-mutate-a-hook}

–•—É–∫–∏ –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –∫–∞–∫ –º–æ–∂–Ω–æ –±–æ–ª–µ–µ ¬´—Å—Ç–∞—Ç–∏—á–Ω—ã–º–∏¬ª. –≠—Ç–æ –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ –∏—Ö –Ω–µ —Å–ª–µ–¥—É–µ—Ç –¥–∏–Ω–∞–º–∏—á–µ—Å–∫–∏ –∏–∑–º–µ–Ω—è—Ç—å. –ù–∞–ø—Ä–∏–º–µ—Ä, —ç—Ç–æ –æ–∑–Ω–∞—á–∞–µ—Ç, —á—Ç–æ –≤—ã –Ω–µ –¥–æ–ª–∂–Ω—ã –ø–∏—Å–∞—Ç—å —Ö—É–∫–∏ –±–æ–ª–µ–µ –≤—ã—Å–æ–∫–æ–≥–æ –ø–æ—Ä—è–¥–∫–∞:

```js hl_lines="2-3"
function ChatInput() {
    // üî¥ Bad: don't write higher order Hooks
    const useDataWithLogging = withLogging(useData);
    const data = useDataWithLogging();
}
```

–•—É–∫–∏ –¥–æ–ª–∂–Ω—ã –±—ã—Ç—å –Ω–µ–∏–∑–º–µ–Ω—è–µ–º—ã–º–∏ –∏ –Ω–µ –ø–æ–¥–≤–µ—Ä–≥–∞—Ç—å—Å—è –º—É—Ç–∞—Ü–∏–∏. –í–º–µ—Å—Ç–æ —Ç–æ–≥–æ —á—Ç–æ–±—ã –¥–∏–Ω–∞–º–∏—á–µ—Å–∫–∏ –∏–∑–º–µ–Ω—è—Ç—å —Ö—É–∫, —Å–æ–∑–¥–∞–π—Ç–µ —Å—Ç–∞—Ç–∏—á–µ—Å–∫—É—é –≤–µ—Ä—Å–∏—é —Ö—É–∫–∞ —Å –Ω—É–∂–Ω–æ–π —Ñ—É–Ω–∫—Ü–∏–æ–Ω–∞–ª—å–Ω–æ—Å—Ç—å—é.

```js hl_lines="2-3 6"
function ChatInput() {
    // ‚úÖ Good: Create a new version of the Hook
    const data = useDataWithLogging();
}

function useDataWithLogging() {
    // ... Create a new version of the Hook and inline the logic here
}
```

### –ù–µ –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ —Ö—É–∫–∏ –¥–∏–Ω–∞–º–∏—á–µ—Å–∫–∏ {#dont-dynamically-use-hooks}

–•—É–∫–∏ —Ç–∞–∫–∂–µ –Ω–µ —Å–ª–µ–¥—É–µ—Ç –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –¥–∏–Ω–∞–º–∏—á–µ—Å–∫–∏: –Ω–∞–ø—Ä–∏–º–µ—Ä, –≤–º–µ—Å—Ç–æ –∏–Ω—ä–µ–∫—Ü–∏–∏ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–µ–π –≤ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–µ –ø–µ—Ä–µ–¥–∞–≤–∞—Ç—å —Ö—É–∫ –≤ –∫–∞—á–µ—Å—Ç–≤–µ –∑–Ω–∞—á–µ–Ω–∏—è:

```js hl_lines="2-3"
function ChatInput() {
    // üî¥ Bad: don't pass Hooks as props
    return <Button useData={useDataWithLogging} />;
}
```

You should always inline the call of the Hook into that component and handle any logic in there.

```js hl_lines="6"
function ChatInput() {
    return <Button />;
}

function Button() {
    const data = useDataWithLogging(); // ‚úÖ Good: Use the Hook directly
}

function useDataWithLogging() {
    // If there's any conditional logic to change the Hook's behavior,
    // it should be inlined into the Hook
}
```

–¢–∞–∫–∏–º –æ–±—Ä–∞–∑–æ–º, `<Button />` –≥–æ—Ä–∞–∑–¥–æ –ø—Ä–æ—â–µ –ø–æ–Ω—è—Ç—å –∏ –æ—Ç–ª–∞–¥–∏—Ç—å. –ö–æ–≥–¥–∞ —Ö—É–∫–∏ –∏—Å–ø–æ–ª—å–∑—É—é—Ç—Å—è –¥–∏–Ω–∞–º–∏—á–µ—Å–∫–∏, —ç—Ç–æ –∑–Ω–∞—á–∏—Ç–µ–ª—å–Ω–æ —É–≤–µ–ª–∏—á–∏–≤–∞–µ—Ç —Å–ª–æ–∂–Ω–æ—Å—Ç—å –≤–∞—à–µ–≥–æ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è –∏ –ø—Ä–µ–ø—è—Ç—Å—Ç–≤—É–µ—Ç –ª–æ–∫–∞–ª—å–Ω–æ–º—É –º—ã—à–ª–µ–Ω–∏—é, –¥–µ–ª–∞—è –≤–∞—à—É –∫–æ–º–∞–Ω–¥—É –º–µ–Ω–µ–µ –ø—Ä–æ–¥—É–∫—Ç–∏–≤–Ω–æ–π –≤ –¥–æ–ª–≥–æ—Å—Ä–æ—á–Ω–æ–π –ø–µ—Ä—Å–ø–µ–∫—Ç–∏–≤–µ. –ö—Ä–æ–º–µ —Ç–æ–≥–æ, —Ç–∞–∫ –ª–µ–≥—á–µ —Å–ª—É—á–∞–π–Ω–æ –Ω–∞—Ä—É—à–∏—Ç—å [–ü—Ä–∞–≤–∏–ª–∞ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏—è —Ö—É–∫–æ–≤](./rules-of-hooks.md), —Å–æ–≥–ª–∞—Å–Ω–æ –∫–æ—Ç–æ—Ä—ã–º —Ö—É–∫–∏ –Ω–µ –¥–æ–ª–∂–Ω—ã –≤—ã–∑—ã–≤–∞—Ç—å—Å—è —É—Å–ª–æ–≤–Ω–æ. –ï—Å–ª–∏ –≤—ã —Å—Ç–æ–ª–∫–Ω—É–ª–∏—Å—å —Å –Ω–µ–æ–±—Ö–æ–¥–∏–º–æ—Å—Ç—å—é –∏–∑–¥–µ–≤–∞—Ç—å—Å—è –Ω–∞–¥ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞–º–∏ –¥–ª—è —Ç–µ—Å—Ç–æ–≤, –ª—É—á—à–µ –≤–º–µ—Å—Ç–æ —ç—Ç–æ–≥–æ –∏–∑–¥–µ–≤–∞—Ç—å—Å—è –Ω–∞–¥ —Å–µ—Ä–≤–µ—Ä–æ–º, —á—Ç–æ–±—ã –æ–Ω –æ—Ç–≤–µ—á–∞–ª –∫–æ–Ω—Å–µ—Ä–≤–∏—Ä–æ–≤–∞–Ω–Ω—ã–º–∏ –¥–∞–Ω–Ω—ã–º–∏. –ï—Å–ª–∏ –µ—Å—Ç—å –≤–æ–∑–º–æ–∂–Ω–æ—Å—Ç—å, –æ–±—ã—á–Ω–æ —ç—Ñ—Ñ–µ–∫—Ç–∏–≤–Ω–µ–µ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞—Ç—å –ø—Ä–∏–ª–æ–∂–µ–Ω–∏–µ —Å –ø–æ–º–æ—â—å—é —Å–∫–≤–æ–∑–Ω—ã—Ö —Ç–µ—Å—Ç–æ–≤.

<small>:material-information-outline: –ò—Å—Ç–æ—á–Ω–∏–∫ &mdash; <https://react.dev/reference/rules/react-calls-components-and-hooks></small>
