---
description: –•—É–∫–∏ –æ–ø—Ä–µ–¥–µ–ª—è—é—Ç—Å—è —Å –ø–æ–º–æ—â—å—é —Ñ—É–Ω–∫—Ü–∏–π JavaScript, –Ω–æ –æ–Ω–∏ –ø—Ä–µ–¥—Å—Ç–∞–≤–ª—è—é—Ç —Å–æ–±–æ–π –æ—Å–æ–±—ã–π —Ç–∏–ø –º–Ω–æ–≥–æ–∫—Ä–∞—Ç–Ω–æ –∏—Å–ø–æ–ª—å–∑—É–µ–º–æ–π –ª–æ–≥–∏–∫–∏ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞ —Å –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏—è–º–∏ –Ω–∞ —Ç–æ, –≥–¥–µ –æ–Ω–∏ –º–æ–≥—É—Ç –±—ã—Ç—å –≤—ã–∑–≤–∞–Ω—ã
---

# –ü—Ä–∞–≤–∏–ª–∞ —Ä–∞–±–æ—Ç—ã —Å —Ö—É–∫–∞–º–∏

<big>
–•—É–∫–∏ –æ–ø—Ä–µ–¥–µ–ª—è—é—Ç—Å—è —Å –ø–æ–º–æ—â—å—é —Ñ—É–Ω–∫—Ü–∏–π JavaScript, –Ω–æ –æ–Ω–∏ –ø—Ä–µ–¥—Å—Ç–∞–≤–ª—è—é—Ç —Å–æ–±–æ–π –æ—Å–æ–±—ã–π —Ç–∏–ø –º–Ω–æ–≥–æ–∫—Ä–∞—Ç–Ω–æ –∏—Å–ø–æ–ª—å–∑—É–µ–º–æ–π –ª–æ–≥–∏–∫–∏ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–∞ —Å –æ–≥—Ä–∞–Ω–∏—á–µ–Ω–∏—è–º–∏ –Ω–∞ —Ç–æ, –≥–¥–µ –æ–Ω–∏ –º–æ–≥—É—Ç –±—ã—Ç—å –≤—ã–∑–≤–∞–Ω—ã.
</big>

## –í—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ —Ç–æ–ª—å–∫–æ –Ω–∞ –≤–µ—Ä—Ö–Ω–µ–º —É—Ä–æ–≤–Ω–µ {#only-call-hooks-at-the-top-level}

–§—É–Ω–∫—Ü–∏–∏, –Ω–∞–∑–≤–∞–Ω–∏—è –∫–æ—Ç–æ—Ä—ã—Ö –Ω–∞—á–∏–Ω–∞—é—Ç—Å—è —Å `use`, –≤ React –Ω–∞–∑—ã–≤–∞—é—Ç—Å—è [_Hooks_](../react/hooks.md).

**–ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ –≤–Ω—É—Ç—Ä–∏ —Ü–∏–∫–ª–æ–≤, —É—Å–ª–æ–≤–∏–π, –≤–ª–æ–∂–µ–Ω–Ω—ã—Ö —Ñ—É–Ω–∫—Ü–∏–π –∏–ª–∏ –±–ª–æ–∫–æ–≤ `try`/`catch`/`finally`.** –í–º–µ—Å—Ç–æ —ç—Ç–æ–≥–æ –≤—Å–µ–≥–¥–∞ –∏—Å–ø–æ–ª—å–∑—É–π—Ç–µ —Ö—É–∫–∏ –Ω–∞ –≤–µ—Ä—Ö–Ω–µ–º —É—Ä–æ–≤–Ω–µ –≤–∞—à–µ–π —Ñ—É–Ω–∫—Ü–∏–∏ React, –ø–µ—Ä–µ–¥ –ª—é–±—ã–º —Ä–∞–Ω–Ω–∏–º –≤–æ–∑–≤—Ä–∞—Ç–æ–º. –í—ã –º–æ–∂–µ—Ç–µ –≤—ã–∑—ã–≤–∞—Ç—å —Ö—É–∫–∏ —Ç–æ–ª—å–∫–æ –≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ —Ñ—É–Ω–∫—Ü–∏–∏ React:

-   ‚úÖ –í—ã–∑—ã–≤–∞–π—Ç–µ –∏—Ö –Ω–∞ –≤–µ—Ä—Ö–Ω–µ–º —É—Ä–æ–≤–Ω–µ –≤ —Ç–µ–ª–µ [—Ñ—É–Ω–∫—Ü–∏–æ–Ω–∞–ª—å–Ω–æ–≥–æ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞](../../learn/your-first-component.md).
-   ‚úÖ –í—ã–∑–æ–≤–∏—Ç–µ –∏—Ö –Ω–∞ –≤–µ—Ä—Ö–Ω–µ–º —É—Ä–æ–≤–Ω–µ –≤ —Ç–µ–ª–µ [–ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–æ–≥–æ —Ö—É–∫–∞](../../learn/reusing-logic-with-custom-hooks.md).

```js
function Counter() {
    // ‚úÖ Good: top-level in a function component
    const [count, setCount] = useState(0);
    // ...
}

function useWindowWidth() {
    // ‚úÖ Good: top-level in a custom Hook
    const [width, setWidth] = useState(window.innerWidth);
    // ...
}
```

–í—ã–∑–æ–≤ —Ö—É–∫–æ–≤ (—Ñ—É–Ω–∫—Ü–∏–π, –Ω–∞—á–∏–Ω–∞—é—â–∏—Ö—Å—è —Å `use`) –≤ –ª—é–±—ã—Ö –¥—Ä—É–≥–∏—Ö —Å–ª—É—á–∞—è—Ö –Ω–µ –ø–æ–¥–¥–µ—Ä–∂–∏–≤–∞–µ—Ç—Å—è, –Ω–∞–ø—Ä–∏–º–µ—Ä:

-   üî¥ –ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ –≤–Ω—É—Ç—Ä–∏ —É—Å–ª–æ–≤–∏–π –∏–ª–∏ —Ü–∏–∫–ª–æ–≤.
-   üî¥ –ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ –ø–æ—Å–ª–µ —É—Å–ª–æ–≤–Ω–æ–≥–æ –æ–ø–µ—Ä–∞—Ç–æ—Ä–∞ `–≤–æ–∑–≤—Ä–∞—Ç–∞`.
-   üî¥ –ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ –≤ –æ–±—Ä–∞–±–æ—Ç—á–∏–∫–∞—Ö —Å–æ–±—ã—Ç–∏–π.
-   üî¥ –ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ –≤ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞—Ö –∫–ª–∞—Å—Å–æ–≤.
-   üî¥ –ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ –≤ —Ñ—É–Ω–∫—Ü–∏—è—Ö, –ø–µ—Ä–µ–¥–∞–≤–∞–µ–º—ã—Ö –≤ `useMemo`, `useReducer` –∏–ª–∏ `useEffect`.
-   üî¥ –ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ –≤–Ω—É—Ç—Ä–∏ –±–ª–æ–∫–æ–≤ `try`/`catch`/`finally`.

–ï—Å–ª–∏ –≤—ã –Ω–∞—Ä—É—à–∏—Ç–µ —ç—Ç–∏ –ø—Ä–∞–≤–∏–ª–∞, –≤—ã –º–æ–∂–µ—Ç–µ —É–≤–∏–¥–µ—Ç—å —ç—Ç—É –æ—à–∏–±–∫—É.

```js
function Bad({ cond }) {
    if (cond) {
        // üî¥ Bad: inside a condition (to fix, move it outside!)
        const theme = useContext(ThemeContext);
    }
    // ...
}

function Bad() {
    for (let i = 0; i < 10; i++) {
        // üî¥ Bad: inside a loop (to fix, move it outside!)
        const theme = useContext(ThemeContext);
    }
    // ...
}

function Bad({ cond }) {
    if (cond) {
        return;
    }
    // üî¥ Bad: after a conditional return (to fix, move it before the return!)
    const theme = useContext(ThemeContext);
    // ...
}

function Bad() {
    function handleClick() {
        // üî¥ Bad: inside an event handler (to fix, move it outside!)
        const theme = useContext(ThemeContext);
    }
    // ...
}

function Bad() {
    const style = useMemo(() => {
        // üî¥ Bad: inside useMemo (to fix, move it outside!)
        const theme = useContext(ThemeContext);
        return createStyle(theme);
    });
    // ...
}

class Bad extends React.Component {
    render() {
        // üî¥ Bad: inside a class component (to fix, write a function component instead of a class!)
        useEffect(() => {});
        // ...
    }
}

function Bad() {
    try {
        // üî¥ Bad: inside try/catch/finally block (to fix, move it outside!)
        const [x, setX] = useState(0);
    } catch {
        const [x, setX] = useState(1);
    }
}
```

–í—ã –º–æ–∂–µ—Ç–µ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –ø–ª–∞–≥–∏–Ω [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks), —á—Ç–æ–±—ã –æ—Ç–ª–æ–≤–∏—Ç—å —ç—Ç–∏ –æ—à–∏–±–∫–∏.

!!!note ""

    [–ü–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–µ —Ö—É–∫–∏](../../learn/reusing-logic-with-custom-hooks.md) _–º–æ–≥—É—Ç_ –≤—ã–∑—ã–≤–∞—Ç—å –¥—Ä—É–≥–∏–µ —Ö—É–∫–∏ (–≤ —ç—Ç–æ–º –∏ –∑–∞–∫–ª—é—á–∞–µ—Ç—Å—è –∏—Ö –Ω–∞–∑–Ω–∞—á–µ–Ω–∏–µ). –≠—Ç–æ —Ä–∞–±–æ—Ç–∞–µ—Ç, –ø–æ—Ç–æ–º—É —á—Ç–æ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏–µ —Ö—É–∫–∏ —Ç–∞–∫–∂–µ –¥–æ–ª–∂–Ω—ã –≤—ã–∑—ã–≤–∞—Ç—å—Å—è —Ç–æ–ª—å–∫–æ –≤–æ –≤—Ä–µ–º—è —Ä–µ–Ω–¥–µ—Ä–∏–Ω–≥–∞ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–∞ —Ñ—É–Ω–∫—Ü–∏–∏.

## –í—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ —Ç–æ–ª—å–∫–æ –∏–∑ —Ñ—É–Ω–∫—Ü–∏–π React {#only-call-hooks-from-react-functions}

–ù–µ –≤—ã–∑—ã–≤–∞–π—Ç–µ —Ö—É–∫–∏ –∏–∑ –æ–±—ã—á–Ω—ã—Ö —Ñ—É–Ω–∫—Ü–∏–π JavaScript. –í–º–µ—Å—Ç–æ —ç—Ç–æ–≥–æ –≤—ã –º–æ–∂–µ—Ç–µ:

‚úÖ –í—ã–∑—ã–≤–∞—Ç—å —Ö—É–∫–∏ –∏–∑ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–æ–≤ —Ñ—É–Ω–∫—Ü–∏–π React.
‚úÖ –í—ã–∑—ã–≤–∞—Ç—å —Ö—É–∫–∏ –∏–∑ [–ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏—Ö —Ö—É–∫–æ–≤](../../learn/reusing-logic-with-custom-hooks.md#extracting-your-own-custom-hook-from-a-component).

–°–ª–µ–¥—É—è —ç—Ç–æ–º—É –ø—Ä–∞–≤–∏–ª—É, –≤—ã –≥–∞—Ä–∞–Ω—Ç–∏—Ä—É–µ—Ç–µ, —á—Ç–æ –≤—Å—è –ª–æ–≥–∏–∫–∞ —Å —Å–æ—Å—Ç–æ—è–Ω–∏–µ–º –≤ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç–µ –±—É–¥–µ—Ç —á–µ—Ç–∫–æ –≤–∏–¥–Ω–∞ –∏–∑ –µ–≥–æ –∏—Å—Ö–æ–¥–Ω–æ–≥–æ –∫–æ–¥–∞.

```js
function FriendList() {
    const [
        onlineStatus,
        setOnlineStatus,
    ] = useOnlineStatus(); // ‚úÖ
}

function setOnlineStatus() {
    // ‚ùå Not a component or custom Hook!
    const [
        onlineStatus,
        setOnlineStatus,
    ] = useOnlineStatus();
}
```

<small>:material-information-outline: –ò—Å—Ç–æ—á–Ω–∏–∫ &mdash; <https://react.dev/reference/rules/rules-of-hooks></small>
